DR-119 dataset pdao

ddietterich opened the pull request:
This PR depends on DR-89. It provides the implementation of the pdao for datasets.

ddietterich committed:
Add dataset flights; passes existing tests; needs tests

ddietterich committed:
Wired up REST API with swagger fixes

ddietterich committed:
Code for dataset flights and test; in debug

ddietterich committed:
Add some logging; fix bugs in create dataset and elsewhere

ddietterich committed:
Rebase to develop

ddietterich committed:
Happy path test works

ddietterich committed:
Fix enumeration test bugs

ddietterich committed:
Remove some needless changes

ddietterich committed:
Findbugs and a test fix

ddietterich committed:
Checkpoint: coded id table load and recursive relationship walk

ddietterich committed:
Checkpoint: create dataset coded; now to test

ddietterich committed:
Checkpoint: code up mapping values to row ids

ddietterich committed:
Revise tests and debug dataset pdao

ddietterich committed:
Added and revised dataset tests, and debugged

ddietterich committed:
...and make checkstyle and findbugs happy

thathert commented on src/main/java/bio/terra/controller/RepositoryApiController.java:
```diff
@@ -100,13 +108,9 @@ protected void initBinder(final WebDataBinder binder) {
-        return new ResponseEntity<>(studySummary, HttpStatus.CREATED);
+    @ExceptionHandler(NotFoundException.class)
+    public ResponseEntity<ErrorModel> handleNotFoundException(NotFoundException ex) {
+        return new ResponseEntity<>(new ErrorModel().message(ex.getMessage()), HttpStatus.BAD_REQUEST);
```
Should this be a `NOT_FOUND` instead of a `BAD_REQUEST`?

ddietterich replied:
Yes, and I changed all of the other not found exceptions to return NOT_FOUND instead of BAD_REQUEST. More fodder for our exception discussion.

thathert commented on src/main/java/bio/terra/flight/dataset/create/CreateDatasetPrimaryDataStep.java:
```diff
@@ -0,0 +1,74 @@
+        DatasetSource source = dataset.getDatasetSources().get(0);
+        RowIdMatch rowIdMatch = bigQueryPdao.mapValuesToRows(dataset, source, contentsModel.getRootValues());
+        if (rowIdMatch.getUnmatchedInputValues().size() != 0) {
+            StringBuilder builder = new StringBuilder("Mismatched input values: ");
```
nit: in Java 8 we can clean this up using String.join:

```java
String unmatchedValues = String.join("', '", rowIdMatch.getUnmatchedInputValues());
String message = String.format("Mismatched input values: '%s'", unmatchedValues);
```

ddietterich replied:
Cool! Done!!

thathert commented on src/main/java/bio/terra/pdao/PrimaryDataAccess.java:
```diff
@@ -50,4 +55,41 @@
+     * @param inputValues
+     * @return RowIdMatch
+     */
+    RowIdMatch mapValuesToRows(bio.terra.metadata.Dataset dataset,
```
Any reason this is fully qualified here?

ddietterich replied:
BigQuery has a class called Dataset, too. I always fully qualify ours. Sometimes I also qualified BigQuery's.

thathert commented on src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java:
```diff
@@ -1,94 +1,249 @@
+import static bio.terra.pdao.PdaoConstant.PDAO_PREFIX;
+import static bio.terra.pdao.PdaoConstant.PDAO_ROW_ID_COLUMN;
+import static bio.terra.pdao.PdaoConstant.PDAO_ROW_ID_TABLE;
+import static bio.terra.pdao.PdaoConstant.PDAO_TABLE_ID_COLUMN;
```
I like this approach, we should do this for constants from now on (and move the JobMapKeys enum over to this style in another PR).

ddietterich replied:
Added this to the Code consistency ticket

thathert commented on src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java:
```diff
@@ -113,6 +268,287 @@ private Schema buildSchema(StudyTable table, boolean addRowIdColumn) {
+        }
+    }
+
+    private void validateRowIdsForRoot(String studyDatasetName,
```
Can you add a comment around what this is validating? It is not clear from the method name.

ddietterich replied:
Done

thathert commented on src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java:
```diff
@@ -113,6 +268,287 @@ private Schema buildSchema(StudyTable table, boolean addRowIdColumn) {
+        String fromColumnName = relationship.getFromColumn().getName();
+
+        /*
+            Building this SQL:
```
Thanks for including these samples before going into the StringBuilder

thathert commented on src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java:
```diff
@@ -113,6 +268,287 @@ private Schema buildSchema(StudyTable table, boolean addRowIdColumn) {
+        try {
+            QueryJobConfiguration queryConfig =
+                    QueryJobConfiguration.newBuilder(sql)
+                            .setDestinationTable(TableId.of(datasetName, PDAO_ROW_ID_TABLE))
```
Very cool.

thathert commented on src/test/java/bio/terra/controller/DatasetOperationTest.java:
```diff
@@ -0,0 +1,341 @@
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(jsonRequest))
+// TODO: swagger field validation errors do not set content type; they log and return nothing
+//                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
```
Can you assert that the response code is a 4xx where we expect validation errors? That's how I've been testing these validations.

ddietterich replied:
I don't expect a failure here. In developing the tests, I had errors in the input file that caused validation errors. But instead of seeing those errors, I was seeing this Expect fail.

I think we need to get the validation errors to produce proper ErrorModel. We can add a common failure check in the validateJobModel and wait.

thathert commented on src/test/java/bio/terra/controller/DatasetOperationTest.java:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    // TODO: this can probably be common code for anything async
+    private MockHttpServletResponse validateJobModelAndWait(MvcResult inResult) throws Exception {
```
Awesome

thathert reviewed (approved):
A couple of style nits and questions, but this looks good to me. I like this approach for walking the relationships.

RoriCremer reviewed (approved):
Probably makes sense to swap out timestamps in a later pr?

ddietterich committed:
Revamped test and fixed some error handling
- taught JobService to notice flight failures and produce better error returns
- taught create service steps to catch errors and return ErrorModels
- add tests using the minimal study
- add a method to load CSVs into study tables from test resources

ddietterich committed:
Add better support for walking and fix infinite recursion

ddietterich committed:
Tidy up. Two failing tests. Going to rebase before fixing them.

ddietterich committed:
Merge from origin/develop

ddietterich committed:
Stomp a few stray bugs

thathert commented on src/main/java/bio/terra/flight/dataset/create/CreateDatasetPrimaryDataStep.java:
```diff
@@ -70,5 +71,14 @@ public StepResult undoStep(FlightContext context) {
     }
+
+    // TODO: this is a copy of what is in the metadata step. Should make a common utility module for
+    // things like this.
```
Can you add something in JIRA to track this? (and possibly other things that would make sense to go in the utility module?)

thathert commented on src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java:
```diff
@@ -342,67 +357,96 @@ private void validateRowIdsForRoot(String studyDatasetName,
+     * the from->to and to->from direction, we have to avoid re-walking a traversed relationship
+     * or we infinite loop. Trust me, I know... :)
+     *
+     * TODO: REVIEWERS: should this code detect circular references?
```
I think that this is the responsibility of the study create validation. We should be able to detect this from the asset spec, tables, and relationship definitions.

I like the pattern of validating as early as possible at the controller layer if we can, and then having the more stateful validation in the flights/daos/etc.

thathert replied:
It would be nice to have in this PR, but I don't think it's a quick couple of lines of code to write. If you feel like it's out of the scope of this work can you add something in the backlog to track?

thathert commented on src/test/java/bio/terra/controller/DatasetOperationTest.java:
```diff
@@ -184,9 +218,77 @@ public void deleteAllBigQueryProjects() throws Exception {
+
+    private void loadStudyData(String studyName, String tableName, String resourcePath) throws Exception {
+        String datasetName = PDAO_PREFIX + studyName;
+        String location = "US"; // ?? Hope this will work. It is what
```
This comment is poetic 😄 I was wondering the same thing when I was testing out the bq cli for ingesting.

thathert reviewed (approved):
Looks good to me
